flocker/acceptance/testtools.py

```diff
@@ -322,133 +326,44 @@ def base_url(self):
-
-        :return: ``Deferred`` firing with a list of dataset dictionaries,
-            the configuration of the cluster.
-        """
-        request = self.treq.get(
-            self.base_url + b"/configuration/datasets", persistent=False)
-        request.addCallback(check_and_decode_json, OK)
-        return request
-
-    @log_method
-    def datasets_state(self):
-        """
-        Return the actual dataset state of the cluster.
+        Poll the dataset state API until the supplied dataset exists.
 
-        :return: ``Deferred`` firing with a list of dataset dictionaries,
-            the state of the cluster.
-        """
-        request = self.treq.get(
-            self.base_url + b"/state/datasets", persistent=False)
-        request.addCallback(check_and_decode_json, OK)
-        return request
+        :param Dataset expected_dataset: The configured dataset that
+            we're waiting for in state.
 
-    @log_method
-    def wait_for_dataset(self, dataset_properties):
+        :returns: A ``Deferred`` which fires with ``expected_datasets``
+            when cluster state matches the configuration for the given
+            dataset.
         """
-        Poll the dataset state API until the supplied dataset exists.
+        expected_dataset_state = DatasetState(
+            dataset_id=expected_dataset.dataset_id,
+            primary=expected_dataset.primary,
+            maximum_size=expected_dataset.maximum_size,
+            path=None)
 
-        :param dict dataset_properties: The attributes of the dataset that
-            we're waiting for.
-        :returns: A ``Deferred`` which fires with an API response when a
-            dataset with the supplied properties appears in the cluster.
-        """
         def created():
             """
             Check the dataset state list for the expected dataset.
             """
-            request = self.datasets_state()
-
-            def got_body(body):
-                # State listing doesn't have metadata or deleted, but does
-                # have unpredictable path.
-                expected_dataset = dataset_properties.copy()
-                del expected_dataset[u"metadata"]
-                del expected_dataset[u"deleted"]
-                for dataset in body:
-                    try:
-                        dataset.pop("path")
-                    except KeyError:
-                        # Non-manifest datasets don't have a path
-                        pass
-                return expected_dataset in body
-            request.addCallback(got_body)
+            request = self.client.list_datasets_state()
+
+            def got_results(results):
+                # State has unpredictable path, so we don't bother
+                # checking for its contents:
+                for dataset in results:
+                    if dataset.set(path=None) == expected_dataset_state:
+                        return True
+                return False
```

Comment:
It took me a while to puzzle the logic of this one. I don't have any good ideas on what can make it more clear. Perhaps

```python
return expected_dataset_state in [dataset.set(path=None) for dataset in results]
```

:question: